### ISSUE TITLE
Reject with an error instead of string when a query mask is set

### ISSUE THREAD

**Issue by madebyherzblut** [vitaly-t/pg-promise#47]
When a query mask (e. g. `one`) is set, the promise is rejected with the string `"No data returned from the query."`. This makes checking for errors very cumbersome and does not play well with the typical express/restify/whatever style:

```javascript
.catch(err => {
  if (typeof err === 'string') return next(new Error(err));
  next(err);
});
```

There are several reasons why you should not [reject with plain strings](http://making.change.org/post/69613524472/promises-and-error-handling). It would be nice to have custom error objects like `UnexpectedResultError`

```javascript
.catch(err => {
  if (err instanceof pgp.UnexpectedResultError) {
    return res.redirect('/somewhere');
  }
  next(err);
});
```

What is your guys opinion on that?

This relates to #37.

**Comment by vitaly-t** [comment]
It can be certainly considered for version 2.0 of the library, when all the breaking changes roll in.

I've just updated the [plan for 2.0](https://github.com/vitaly-t/pg-promise/wiki/2.0).

In the meantime, you can use a more elegant way to check for the error message:

```javascript
.catch(err)
{
    var msg = err.message || err;
}
```

**Comment by vitaly-t** [comment]
@madebyherzblut Reading it once more, I still do not see relevance to #37. That one is about checking the row being found or not, which isn't an error at all, and you were talking about the errors support here.

**Comment by madebyherzblut** [comment]
For me both issues are about the same problem: I consider everything that rejects my promise as an error and right now checking for an error is awkward because sometimes  you get an `Error` object and sometimes a string:

> …if it's a real syntax error I get a complex object, but for no data found it's just a string saying "No data returned from the query.". It would be nicer if there were a error object with a code and message for no data found.

This is how I do it now:

```javascript
db.one(query).catch(err => {
  if (typeof err === 'string') {
    // This must be a pgp error caused by a mismatched query result mask
    return next(new error.NotFoundError(err));
  }
  // This is likely a database error
  next(new error.InternalServerError(err));
});
```

**Comment by vitaly-t** [comment]
That's the thing, the issue you referred to isn't an error, and it doesn't reject. That was about use of method `oneOrNone` that resolves with `null` when no record is found. What you are talking about is different.

**Comment by madebyherzblut** [comment]
Ah, sorry you are right. I somehow had `one` instead of `oneOrNone` in my head.

**Comment by vitaly-t** [comment]
This has been implemented within the 2.0 branch (currently pre-release).

Type `QueryResultError` is returned when the returned dataset doesn't match the expected one according to the Query Result Mask.

**Comment by vitaly-t** [comment]
@madebyherzblut In case you are still using the library, upgrading to the latest version will require a change in using type [QueryResultError](http://vitaly-t.github.io/pg-promise/QueryResultError.html):

It is no longer available directly from the root, it is available from namespace `errors`, i.e. as `errors.QueryResultError`, as explained in the API ;)
